homeassistant/components/snooz/__init__.py

```diff
@@ -60,3 +61,103 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
             hass.data.pop(DOMAIN)
 
     return unload_ok
+
+
+async def _async_load_entry_data(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> SnoozConfigurationData:
+    if _entry_missing_device_config(entry):
+        adv_data = await _async_update_entry_config(hass, entry)
+    else:
+        adv_data = _load_entry_config(entry)
+
+    address: str = entry.data[CONF_ADDRESS]
+    if not (ble_device := async_ble_device_from_address(hass, address)):
+        raise ConfigEntryNotReady(
+            f"Could not find Snooz with address {address}. Try power cycling the device"
+        )
+
+    return SnoozConfigurationData(
+        ble_device,
+        adv_data,
+        device=SnoozDevice(ble_device, adv_data),
+        title=entry.title,
+    )
+
+
+def _entry_missing_device_config(entry: ConfigEntry) -> bool:
+    return CONF_MODEL not in entry.data or CONF_FIRMWARE_VERSION not in entry.data
+
+
+def _load_entry_config(entry: ConfigEntry) -> SnoozAdvertisementData:
+    return SnoozAdvertisementData(
+        model=SnoozDeviceModel(entry.data[CONF_MODEL]),
+        firmware_version=SnoozFirmwareVersion(entry.data[CONF_FIRMWARE_VERSION]),
+        password=entry.data[CONF_TOKEN],
+    )
+
+
+async def _async_update_entry_config(
+    hass: HomeAssistant, entry: ConfigEntry
+) -> SnoozAdvertisementData:
+    """Find a supported advertisement and store data in config entry."""
+
+    address = entry.data[CONF_ADDRESS]
+    LOGGER.debug(f"Loading device information from advertisement for address {address}")
+
+    adv_data = await hass.async_create_task(
+        _async_wait_for_advertisement(hass, address)
+    )
```

Comment:
Can you split this function up and or rename it to better show intent. Its doing three things right now.

1. Updating the config entry
2. Waiting for an advertisement
3. Optionally extending the adv data from the config entry data